### PR TITLE
🚀 Support offline players in ExprBed

### PR DESCRIPTION
**Pull request by AyhamAl-Ali**
### Description

- Due to API limitations, OfflinePlayer class doesn't have `setBedSpawnLocation` method therefore it has been excluded in the acceptChange method
- Added the ability to set **safe/valid** bed location so that it will become invalid if the location has been obstructed or doesn't exist anymore. (unsafe/invalid i.e. force set is still the default behavior)

---
**Target Minecraft Versions:** Any
**Requirements:** None
**Related Issues:**  
- #4659

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprBed.java [hunk 1]
```diff
@@ -30,46 +32,47 @@
...
-		} else {
-			final Location l = (Location) delta[0];
-			for (final Player p : getExpr().getArray(e)) {
-				p.setBedSpawnLocation(l, true);
-			}
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		Location loc = delta == null ? null : (Location) delta[0];
+		for (OfflinePlayer p : getExpr().getArray(e)) {
+			if (p.isOnline()) // double check
+				((Player) p).setBedSpawnLocation(loc, true);
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprBed.java:75]: should `force` be `true`? the docs for the expression say "the bed still exists and is unobstructed", which could potentially be false if set manually. It looks like this was the existing behavior though so maybe it's juts a documentation fix that's needed?

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprBed.java:75]: Yep, seems like it was [added ](https://github.com/SkriptLang/Skript/commit/d0e49bd57828f240cac19c6bd669bff102bfdf83#diff-0bea3bd93975380fd06a01d409b2e5bf4d1aba00bf11ec50c4b7f9e4da3a6bbfR69)the acceptChange method later without changing the description, will do it 👍

File: src/main/java/ch/njol/skript/expressions/ExprBed.java [hunk 2]
```diff
@@ -30,46 +36,61 @@
...
+		register(ExprBed.class, Location.class, "[(:(safe|valid)|:(unsafe|invalid))] bed[s] [location[s]]", "offlineplayers");
 	}
-	
+
+	private boolean isSafe = false; // default behavior is unsafe as used to be
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		isSafe = parseResult.hasTag("safe") || parseResult.hasTag("valid");
+		System.out.println("TAGS: " + parseResult.tags);
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprBed.java:64]: think you left this in on accident

**Review by Pikachu920 — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprBed.java (hunk 2 above)
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprBed.java:64]: Ops🤦‍♂️

**Review by Pikachu920 — APPROVED** [review]

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprBed.java [hunk 3]
```diff
@@ -30,46 +36,60 @@
...
+	"The bed location of a player, " +
+	"i.e. the spawn point of a player if they ever slept in a bed and the bed still exists and is unobstructed however, " +
+	"you can set the unsafe bed location of players and they will respawn there even if it has been obstructed or doesn't exist anymore.",
+	"Note that when using unsafe/invalid bed location syntax that will not affect the returned value, it only affects when setting it."
+})
+@Examples({
+	"if bed of player exists:",
+		"\tteleport player the the player's bed",
+	"else:",
+		"\tteleport the player to the world's spawn point"
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprBed.java:50]:
```suggestion
	"if bed of player exists:",
	"\tteleport player the the player's bed",
	"else:",
	"\tteleport the player to the world's spawn point"
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Just this + TP's comments

**Comment by TheLimeGlass** [comment]
I don't understand what the point of this pull request is when they still have to be online for their bed location to be set anyways, why make it offline player?

File: src/main/java/ch/njol/skript/expressions/ExprBed.java [hunk 4]
```diff
@@ -30,46 +36,61 @@
...
 	
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.SET || mode == ChangeMode.DELETE)
-			return new Class[] {Location.class};
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.DELETE) {
+			if (!Player.class.isAssignableFrom(getExpr().getReturnType())) {
+				Skript.error("Bed location of offline players cannot be set/deleted.");
```
[the remark this answers is not in the document]
Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/expressions/ExprBed.java:80]: IMO your suggestion makes it more confusing, a `player` for users doesn't mean  an `online` player, it just means any player, while saying offline players it much clearer that you need an online player for this expression

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprBed.java [hunk 5]
```diff
@@ -18,58 +18,76 @@
...
-			final Location l = (Location) delta[0];
-			for (final Player p : getExpr().getArray(e)) {
-				p.setBedSpawnLocation(l, true);
-			}
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		Location loc = delta == null ? null : (Location) delta[0];
+		for (OfflinePlayer p : getExpr().getArray(e)) {
+			Player op = p.getPlayer();
+			if (op != null) // is online
+				p.getPlayer().setBedSpawnLocation(loc, !isSafe);
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/expressions/ExprBed.java:90]:
```suggestion
				op.setBedSpawnLocation(loc, !isSafe);
```

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**